File: http-streams/src/main/scala/http/PathApi.scala
package sc.http

import akka.http.scaladsl.server.util.Tuple
import akka.http.scaladsl.server.{Directive, PathMatcher}

abstract class PathApi {
  //#pathApi1
  def path[L](pm: PathMatcher[L]): Directive[L]
  //#pathApi1
  //#pathApi
  abstract class PathMatcher[L: Tuple]
  type PathMatcher0 = PathMatcher[Unit]
  type PathMatcher1[T] = PathMatcher[Tuple1[T]]
  type PathMatcher2[T,U] = PathMatcher[Tuple2[T,U]]
  // .. etc
  //#pathApi
}

File: http-streams/src/main/java/http/PathApi.java
package http;

import akka.http.javadsl.server.PathMatcher0;
import akka.http.javadsl.server.PathMatcher1;
import akka.http.javadsl.server.PathMatcher2;
import akka.http.javadsl.server.Route;

import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Supplier;

/**
 * Show what the path directives API looks like in Java interface methods
 */
public class PathApi {
  //#pathApi1
  Route path(String segment, Supplier<Route> inner) {
    //#pathApi1
    return null;
  }

  //#pathApi
           Route path(PathMatcher0 p, Supplier<Route> inner) {
    //#pathApi
    return null;
  }

  //#pathApi
       <T> Route path(PathMatcher1<T> p, Function<T, Route> inner) {
    //#pathApi
    return null;
  }

  //#pathApi
  <T1, T2> Route path(PathMatcher2<T1, T2> p, BiFunction<T1, T2, Route> inner) {
    //#pathApi
    return null;
  }

}
